misc/list_important_dependencies.py

```diff
@@ -0,0 +1,364 @@
+#!/usr/bin/env python3
+# Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+"""
+List versions of what we consider important dependencies of our client & server application.
+"""
+
+from __future__ import annotations
+
+import argparse
+import csv
+import tomllib
+from abc import ABC, abstractmethod
+from collections.abc import Callable, Generator, Sequence
+from enum import Enum, auto
+from pathlib import Path
+from typing import BinaryIO, NamedTuple
+
+from pydantic import BaseModel, ConfigDict, Field, RootModel
+from pydantic.dataclasses import dataclass
+
+ROOT_DIR = Path(__file__).parent.parent.resolve()
+NPM_LOCK_FILES = (
+    ROOT_DIR / "client/package-lock.json",
+    ROOT_DIR / "client/electron/package-lock.json",
+)
+CARGO_LOCK_FILE = ROOT_DIR / "Cargo.lock"
+# We do not include the lock file for the documentations as it's not important for the application.
+PYTHON_LOCK_FILE = ROOT_DIR / "server/poetry.lock"
+RUST_TOOLCHAIN_FILE = ROOT_DIR / "rust-toolchain.toml"
+VERSION_FILE = ROOT_DIR / "misc/versions.toml"
+
+
+@dataclass(frozen=True, slots=True)
+class Dependency:
+    name: str
+    version: str
+    type: DependencyType
+
+
+class DependencyType(Enum):
+    Rust = auto()
+    Python = auto()
+    Javascript = auto()
+
+    def __str__(self) -> str:
+        return self.name
+
+
+def main():
+    args = parse_args()
+
+    with open(VERSION_FILE, "rb") as versions_file:
+        versions: dict[str, str] = tomllib.load(versions_file)
+
+    with open(args.output, "w") as csvfile:
+        csv_writer = csv.DictWriter(
+            csvfile,
+            fieldnames=Dependency.__pydantic_fields__.keys(),  # pyright: ignore [reportAttributeAccessIssue]
+            strict=True,
+            dialect="unix",
+        )
+        csv_writer.writeheader()
+
+        if not args.skip_javascript:
+            deps = list_javascript_deps(versions)
+            write_deps(csv_writer, deps)
+
+        if not args.skip_rust:
+            deps = list_rust_deps(versions)
+            write_deps(csv_writer, deps)
+
+        if not args.skip_python:
+            deps = list_python_deps(versions)
+            write_deps(csv_writer, deps)
+
+
+DependencyModel = RootModel[Dependency]
+
+
+def write_deps(csv: csv.DictWriter[Dependency], deps: set[Dependency]):
+    deps_list = list(deps)
+    deps_list.sort(key=lambda dep: dep.name)
+    csv.writerows(DependencyModel(dep).model_dump() for dep in deps_list)
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("--output", default=Path("versions.csv"), type=Path)
```

Comment:
nitpick: default output file name is not displayed when using `--help`, nor is printed when running the command:
```shell
$ python misc/list_important_dependencies.py
Loading client/package-lock.json content
Loading client/electron/package-lock.json content
Loading rust-toolchain.toml content
Loading Cargo.lock content
Loading server/poetry.lock content
```

Since you have already put some `Loading .... content` prints, I think the consistent thing to do is to add a final `Writting versions CSV ./output.csv` print